zrml/global-disputes/src/lib.rs

```diff
@@ -243,13 +283,76 @@ mod pallet {
                 ExistenceRequirement::AllowDeath,
             )?;
 
+            Self::update_winner(&market_id, &outcome, voting_outcome_fee);
+            let possession =
+                Some(Possession::Paid { owner: owner.clone(), fee: voting_outcome_fee });
+            let outcome_info = OutcomeInfo { outcome_sum: voting_outcome_fee, possession };
+            <Outcomes<T>>::insert(market_id, outcome.clone(), outcome_info);
+
             Self::deposit_event(Event::AddedVotingOutcome { market_id, owner, outcome });
             // charge weight for successfully have no owners in Winners
             // this is the case, because owners are not inserted
             // as long as the global dispute is not finished
             Ok((Some(T::WeightInfo::add_vote_outcome(0u32))).into())
         }
 
+        /// Return the voting outcome fees in case the global dispute was destroyed.
+        ///
+        /// # Arguments
+        ///
+        /// - `market_id`: The id of the market.
+        ///
+        /// # Weight
+        ///
+        /// Complexity: `O(n)`,
+        /// where `n` is the number of all existing outcomes for a global dispute.
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::refund_vote_fees(
+            T::RemoveKeysLimit::get(),
+            T::MaxOwners::get()
+        ))]
+        pub fn refund_vote_fees(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+
+            let gd_info = <GlobalDisputesInfo<T>>::get(market_id)
+                .ok_or(Error::<T>::NoGlobalDisputeInitialized)?;
+            ensure!(gd_info.status == GDStatus::Destroyed, Error::<T>::UnfinishedGlobalDispute);
+
+            let mut owners_len = 0u32;
+            let mut removed_keys_amount = 0u32;
+            for (_, outcome_info) in
+                <Outcomes<T>>::drain_prefix(market_id).take(T::RemoveKeysLimit::get() as usize)
```

Comment:
I think it is better to have a test for this specially for the case where there is more than one iteration of refund_vote_fees is required. To me it feels that drain_prefix removes all items with k1 in one shot.

Reply:
I assume that OutcomesPartiallyCleaned event requires user to execute the extrinsic again. Correct if this is wrong.

Reply:
This is correct. In case there are more than `RemoveKeysLimit` storage values behind the key `market_id` the extrinsic will stop and emit the event `OutcomesPartiallyCleaned`. This requires at least one more execution of this extrinsic.

Reply:
> I think it is better to have a test for this specially for the case where there is more than one iteration of refund_vote_fees is required

Good point. I will fix this.

Reply:
I think there is similar code in purge_outcome or similar method. Please consider them too when you fix this.

Reply:
6ea85bc41839b7ab731b50840637bb42b5e99522 purge_outcomes partially cleaned test

Reply:
1a2c491f7290efde9c3a30a99eea8dbb1f2d876a refund_vote_fees